server/server.go

```diff
@@ -0,0 +1,366 @@
+}
+
+func RefValueToExprValue(res ref.Value) (*eval.ExprValue, error) {
+	if types.IsError(res) {
+		return &eval.ExprValue{
+			Kind: &eval.ExprValue_Error{}}, nil
+	}
+	if types.IsUnknown(res) {
+		return &eval.ExprValue{
+			Kind: &eval.ExprValue_Unknown{}}, nil
+	}
+	v, err := RefValueToValue(res)
+	if err != nil {
+		return nil, err
+	}
+	return &eval.ExprValue{
+		Kind: &eval.ExprValue_Value{Value: v}}, nil
+}
+
+// Convert res, which must not be error or unknown, to a Value proto.
+func RefValueToValue(res ref.Value) (*value.Value, error) {
+	switch res.Type() {
+	case types.BoolType:
+		return &value.Value{
+			Kind: &value.Value_BoolValue{res.Value().(bool)}}, nil
+	case types.BytesType:
+		return &value.Value{
+			Kind: &value.Value_BytesValue{res.Value().([]byte)}}, nil
+	case types.DoubleType:
+		return &value.Value{
+			Kind: &value.Value_DoubleValue{res.Value().(float64)}}, nil
+	case types.IntType:
+		return &value.Value{
+			Kind: &value.Value_Int64Value{res.Value().(int64)}}, nil
+	case types.ListType:
+		l := res.Value().(traits.Lister)
+		sz := l.Size().Value().(int64)
+		elts := make([]*value.Value, sz)
+		for i := int64(0); i < sz; i++ {
+			v, err := RefValueToValue(l.Get(types.Int(i)))
+			if err != nil {
+				return nil, err
+			}
+			elts = append(elts, v)
+		}
+		return &value.Value{
+			Kind: &value.Value_ListValue{
+				&value.ListValue{Values: elts}}}, nil
+	case types.MapType:
+		mapper := res.Value().(traits.Mapper)
+		sz := mapper.Size().Value().(int64)
+		entries := make([]*value.MapValue_Entry, sz)
+		it := mapper.Iterator()
+		for it.HasNext().(types.Bool) {
+			k := it.Next()
+			v := mapper.Get(k)
+			kv, err := RefValueToValue(k)
+			if err != nil {
+				return nil, err
+			}
+			vv, err := RefValueToValue(v)
+			if err != nil {
+				return nil, err
+			}
+			entries = append(entries, &value.MapValue_Entry{Key: kv, Value: vv})
+		}
+		return &value.Value{
+			Kind: &value.Value_MapValue{
+				&value.MapValue{Entries: entries}}}, nil
+	case types.NullType:
+		return &value.Value{
+			Kind: &value.Value_NullValue{}}, nil
+	case types.StringType:
+		return &value.Value{
+			Kind: &value.Value_StringValue{res.Value().(string)}}, nil
+	case types.TypeType:
+		typeName := res.Value().(string)
+		var basicType value.TypeValue_BasicType
+		basicType = value.TypeValue_BASIC_TYPE_UNSPECIFIED
+		switch typeName {
+		case "bool":
+			basicType = value.TypeValue_BOOL_TYPE
+		case "bytes":
+			basicType = value.TypeValue_BYTES_TYPE
+		case "double":
+			basicType = value.TypeValue_DOUBLE_TYPE
+		case "null_type":
+			basicType = value.TypeValue_NULL_TYPE
+		case "int":
+			basicType = value.TypeValue_INT_TYPE
+		case "list":
+			basicType = value.TypeValue_LIST_TYPE
+		case "map":
+			basicType = value.TypeValue_MAP_TYPE
+		case "string":
+			basicType = value.TypeValue_STRING_TYPE
+		case "type":
+			basicType = value.TypeValue_TYPE_TYPE
+		case "uint":
+			basicType = value.TypeValue_UINT_TYPE
+		}
```

Comment:
Consider making a constant map for the above.

Reply:
Done, and for the reverse conversion.